### PR TITLE
feat: add recursive text chunker (fixes #3)

### LINKED ISSUE #3: Recursive / Semantic Text Chunker
## Description

Add a `RecursiveChunker` that splits text using a priority list of separators (`\n\n`, `\n`, `. `, ` `) — falling back to the next separator when a chunk is still too long.

## Motivation

The existing `FixedSizeChunker` splits text at arbitrary character boundaries, which can break mid-sentence or mid-paragraph and degrade retrieval quality. A recursive approach preserves semantic boundaries.

## Acceptance criteria

- [ ] Class `RecursiveChunker` in `ragframework/document/chunkers.py`
- [ ] Subclasses `TextChunker` from `ragframework/base.py`
- [ ] Configurable `separators: list[str]`, `chunk_size: int`, `chunk_overlap: int`
- [ ] Falls back through separator list until chunks fit within `chunk_size`
- [ ] Unit tests covering: long word (no separator fits), empty doc, single-separator doc, overlap behaviour
- [ ] Exported from `ragframework/document/__init__.py`
- [ ] `CHANGELOG.md` updated under `[Unreleased]`

## Files to touch

- `ragframework/document/chunkers.py` — add `RecursiveChunker`
- `ragframework/document/__init__.py` — export it
- `tests/test_document/test_chunkers.py` — add tests

## Resources

- LangChain's `RecursiveCharacterTextSplitter` for inspiration
- Existing chunkers in `ragframework/document/chunkers.py` as reference

### PR DESCRIPTION
**Pull request by Sahvendra7**
## Summary

Implement `RecursiveChunker` for Issue #3 with recursive structural splitting, bounded overlap, and character-level fallback.

## What Changed

- Added `RecursiveChunker` to `ragframework.document.chunkers`
- Added recursive separator hierarchy:
  paragraphs → lines → sentences → words → characters
- Added hard chunk-size enforcement
- Added segment-aware bounded overlap
- Added character-level fallback for oversized text
- Added public export in `ragframework/document/__init__.py`
- Updated `CHANGELOG.md`
- Added unit tests covering required behavior and edge cases

## Correctness

- Empty documents return `[]`.
- Every output chunk respects the configured `chunk_size`.
- Overlap never exceeds the configured `chunk_overlap`.
- With `chunk_overlap=0`, chunk contents reconstruct the original document exactly.
- Document metadata is copied rather than mutated.
- Chunk IDs follow the existing deterministic convention.

## Validation

- `pytest tests/` — 43/43 tests passed
- `ruff check .` — passed
- `black --check .` — passed
- `isort --check-only .` — passed
- `mypy ragframework/` — passed

## Scope

Existing `FixedSizeChunker` and `SentenceChunker` implementations were left unchanged. No external dependencies or unrelated framework changes were introduced.

Fixes #3

### CODE REVIEW

File: ragframework/document/chunkers.py [hunk 1]
```diff
@@ -96,3 +96,147 @@ def chunk(self, document: Document) -> list[Chunk]:
...
+                    break
+
+        if separator == "":
+            result = []
+            step = self.chunk_size - self.chunk_overlap
+            idx = 0
+            while idx < len(text):
+                end = min(idx + self.chunk_size, len(text))
+                result.append(text[idx:end])
+                idx += step
```
Comment (adaumsilva) [inline ragframework/document/chunkers.py:174]: The character-level fallback continues creating chunks after a chunk has already reached the end of the text.

For example, with `text="x" * 10`, `chunk_size=6`, and `chunk_overlap=4`, it produces chunk lengths `[6, 6, 6, 4, 2]`. The last two chunks contain no new content, which can create unnecessary embeddings, storage usage, and duplicate retrieval results.

Please stop the loop when `end == len(text)` and add a regression test for this boundary condition.

doc = Document(id="x", content="x" * 10, metadata={})
chunks = RecursiveChunker(
    separators=[],
    chunk_size=6,
    chunk_overlap=4,
).chunk(doc)

assert [len(c.content) for c in chunks] == [6, 6, 6]

**Review by adaumsilva — CHANGES_REQUESTED** [review]
The implementation is focused and generally satisfies Issue #3. However, the character-level fallback generates redundant trailing chunks when overlap is enabled. Please fix this boundary condition and add a regression test before merging.

**Comment by adaumsilva** [comment]
Thanks for the contribution @Sahvendra7 ! I’ve submitted a review with one boundary-case issue involving redundant chunks when overlap is enabled. Once that is fixed and covered by a regression test, I’ll review it again.

**Review by adaumsilva — APPROVED** [review]
Thanks for addressing the requested change. The character-level fallback now stops correctly after reaching the end of the text, and the regression test covers the reported boundary case. I verified the full test suite and quality checks successfully. Approved.
